lib/Effect/Fs.hs

```diff
@@ -0,0 +1,174 @@
+-- SPDX-FileCopyrightText: 2022 Serokell <https://serokell.io>
+--
+-- SPDX-License-Identifier: MPL-2.0
+
+{-# LANGUAGE ImportQualifiedPost #-}
+module Effect.Fs
+  ( FsEffect
+  , nodeExists
+  , getNode
+  , listDirectory
+  , listDirectoryRec
+  , runFsInIO
+  , stringToPath
+  , pathToString
+  , FsError
+  , Node
+  , Node'
+  , File (..)
+  , Directory (..)
+  )
+  where
+
+import Control.Lens
+import Control.Monad (forM)
+import Data.ByteString (ByteString)
+import Data.Text qualified as T
+import Data.Text.Encoding (decodeUtf8', encodeUtf8)
+import Polysemy
+import Polysemy.Error
+import System.Directory hiding (listDirectory)
+import System.Directory qualified as D
+
+type Node f d = Either (File f) (Directory d)
+type Node' a = Node a a
+type Path = ByteString
+
+newtype File a = File a
+  deriving stock (Show)
+newtype Directory a = Directory a
+  deriving stock (Show)
+
+data FsError
+  = FENodeNotFound Path
+  | FENodeExists (MismatchError (Node' Path))
+  | FEMissingParent (Node' Path)
+  | FEInvalidPath Path
+  deriving stock (Show)
+
+data MismatchError a = MismatchError
+  { meWanted :: a
+  , mwFound :: a
+  }
+  deriving stock (Show)
+
+newtype NodeRec = NodeRec (Node Path [NodeRec])
+
+pathToString :: Path -> Either FsError String
+pathToString path =
+  case decodeUtf8' path of
+    Left a -> Left $ FEInvalidPath path
+    Right b -> Right $ T.unpack b
+stringToPath :: String -> Path
+stringToPath = encodeUtf8 . T.pack
+
+extractNodePath
+  :: Node' Path
+  -> Path
+extractNodePath =
+  \case Left (File path) -> path
+        Right (Directory path) -> path
+
+eitherError
+  :: Member (Error e) r
+  => (a -> e)
+  -> Either a b
+  -> Sem r b
+eitherError f = either (throw . f) pure
+
+data FsEffect m a where
+  NodeExists :: Path -> FsEffect m (Maybe (Node' ()))
+  GetNode :: Path -> FsEffect m (Maybe (Node' Path))
+  ListDirectory :: Directory Path -> FsEffect m [Node' ByteString]
```

Comment:
Shouldn't `ListDirectory` return a `Maybe`, and `Nothing` when the given path doesn't exist?